### PR TITLE
components: Promote VisuallyHidden from ui into full components

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Promotes `ui/visually-hidden` into `components/src/visually-hidden` and exports it as the canonical VisuallyHidden for wp/components.

Please test this PR, test out a few places that use VisuallyHidden 🙂 They have the exact same API so there should be no differences between the two but just in case!

## How has this been tested?
Storybook, unit tests

## Types of changes
New feature! This actually expands the API of VisuallyHidden to have a properly implemented `as` and polymorphism. Not sure if we should include that in the CHANGELOG.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/visually-hidden/component.js [hunk 1]
```diff
@@ -10,12 +10,12 @@ import { useVisuallyHidden } from './hook';
...
+ * import { VisuallyHidden } from `@wordpress/components`;
  *
  * function Example() {
  * 	return (
  * 		<VisuallyHidden>
- * 			<View as="label">Code is Poetry</View>
+ * 			<label>Code is Poetry</label>
  * 		</VisuallyHidden>
  * 	);
  * }
```
Comment (youknowriad) [inline packages/components/src/visually-hidden/component.js:21]: I just noticed this `createComponent` util. Looking at its code, it seems it doesn't do much and basically coping that component here would make things simpler for me at least.

What do you think about it? Is it a good abstraction to have?

Comment (youknowriad) [reply packages/components/src/visually-hidden/component.js:21]: For instance why `useVisuallyHidden` is a hook in the first place, it just adds a class, so why not just add it inline?

Comment (sarayourfriend) [reply packages/components/src/visually-hidden/component.js:21]: I think it's a good abstraction to have and encourages "pure hook" components, which I think is generally a good thing (makes us think twice before using `Context` or adding extra components to the React tree for example).

Additionally it's used for a _ton_ of components in `ui` and refactoring them all to just copy the code from `createComponent` would increase the maintenance cost of all those components.

Comment (youknowriad) [reply packages/components/src/visually-hidden/component.js:21]: > makes us think twice before using Context or adding extra components to the React tree for example

Why is this a bad thing?

Comment (sarayourfriend) [reply packages/components/src/visually-hidden/component.js:21]: Additionally, it enforces the use of `View` where we will be able to add back the base styles that we lost by moving away from the custom style system. Because everything is ultimately a View we can just apply the base styles there.

Comment (sarayourfriend) [reply packages/components/src/visually-hidden/component.js:21]: > Why is this a bad thing?

We should avoid adding extra components to the React tree as much as possible right? I thought that was a general goal in Gutenberg as a performance consideration? Especially in wp/components where the components are used in a wide range of places, every component added to the tree could add up across the app.

React `Context` providers come with a performance cost as well as adding more components to the tree, right?

Maybe I'm mistaken about this 🤷‍♀️

Comment (sarayourfriend) [reply packages/components/src/visually-hidden/component.js:21]: Generally I like the `createComponent` function. Reakit uses a similar abstraction, FWIW, so it's not completely unfounded.

Comment (youknowriad) [reply packages/components/src/visually-hidden/component.js:21]: > We should avoid adding extra components to the React tree as much as possible right? I thought that was a general goal in Gutenberg as a performance consideration? 

Curious where this guideline comes from :P I don't think this was ever a rule in Gutenberg and I don't really understand how adding components to the tree causes performance regressions.

> React Context providers come with a performance cost as well as adding more components to the tree, right?

Depending how it's used, I can see that yes, and I do agree that Context is generally something not very useful for UI components but I also don't think there's any hard rule there.

Comment (youknowriad) [reply packages/components/src/visually-hidden/component.js:21]: Now for the createComponent abstraction itself, to be honest, I personally don't like, it's like we're inventing a new way to write React components and feel like it's just a cosmetic abstraction. 

That said, it's internal to the ui components package, so I'm fine keeping it for now. My main worry is that it's an abstraction that doesn't bring much while raising the bar for casual components to understand how components are built in this package, why it's just just a regular component.

Comment (sarayourfriend) [reply packages/components/src/visually-hidden/component.js:21]: Well, I think I just picked it up from discussions around the repository and in the WordPress slack 🤔 But maybe I'm imagining things lol! I feel like I remember @gwwar or maybe @ellatrix mentioning something about trying to remove components from the React component tree?

Comment (youknowriad) [reply packages/components/src/visually-hidden/component.js:21]: Now looking more at createCompnent, It does call `memo` by default in all components, that personally doesn't make sense to me. It's going to do the opposite of what it's meant to solve. Basically, most of these small components don't need to be memoized, React already only update the underlying Dom element if its prop changes, so this extra `memo` is useless. At least we should change the default value to false there.

Comment (sarayourfriend) [reply packages/components/src/visually-hidden/component.js:21]: > My main worry is that it's an abstraction that doesn't bring much while raising the bar for casual components to understand how components are built in this package, why it's just just a regular component.

That's fair, it is a new abstraction and does technically redirect from the core functionality of a component.

How do you envision, for example, the `Text` component looking if we removed `createComponent`? Would we get rid of the `useText` hook?

My concern is that if we remove `createComponent`, we'll just be duplicating a ton of code everywhere, indeed, that's why Reakit uses it for example. If a component is follows the pattern established in `createComponent`, namely:

1. Push all props through a hook
2. Pass all processed props to View
3. Connect to the contextSystem

Then we've just duplicated that code, well, more than duplicated, we've replicated probably dozens of times over across the components codebase, just to avoid a tidy, well-tested abstraction. Is that worth it? 🤷‍♀️

Comment (youknowriad) [reply packages/components/src/visually-hidden/component.js:21]: > How do you envision, for example, the Text component looking if we removed createComponent? Would we get rid of the useText hook?

Well!  for me these `useVisuallyHidden`, `useText`... hooks are very weird hooks. They don't really match what hooks are meant to be used for. They correspond to "components" and not "behavior". It looks like a hack to the hooks system :P 

> Push all props through a hook

What is this exactly? :P Is it a requirement to pass all props through a hook?

> Push all props through a hook

Do all components need to support "as" and other features proposed by the "View" component? I guess my opinion is probably "no" Why should an `Icon` component support that? an Icon always renders an svg (just an example)

> Connect to the contextSystem

Do all components need to connect to the context system. Isn't that just useless overhead for most components? Not sure I understand that system well enough but it seems only useful for components that compose (Card, CardBody...)

Comment (sarayourfriend) [reply packages/components/src/visually-hidden/component.js:21]: > At least we should change the default value to false there.

Agreed! We need to update that in `contextConnect` as well.

File: packages/components/src/visually-hidden/hook.js [hunk 2]
```diff
@@ -8,11 +8,8 @@ import { cx } from 'emotion';
...
 
-// duplicate this for the sake of being able to export it, it'll be removed when we replace VisuallyHidden in components/src anyway
-/** @typedef {import('../context').ViewOwnProps<{}, 'div'>} Props */
-
 /**
- * @param {import('../context').ViewOwnProps<{}, 'div'>} props
+ * @param {import('../ui/context').ViewOwnProps<{}, 'div'>} props
  */
 export function useVisuallyHidden( { className, ...props } ) {
 	// circumvent the context system and write the classnames ourselves
```
Comment (youknowriad) [inline packages/components/src/visually-hidden/hook.js:15]: should we remove the `wp-components-visually-hidden` classname?

File: packages/components/src/visually-hidden/style.scss [hunk 3]
```diff
@@ -1,30 +0,0 @@
-.components-visually-hidden {
```
Comment (youknowriad) [inline packages/components/src/visually-hidden/style.scss:1]: Removing all these styles could potentially impact some third-party plugins that rely on them. The good news is that these are not official APIs, meaning plugins should be using components and not classnames. I think once we make more progress into merging G2 components and removing these old styles, we should consider writing a dev note to document the removed classnames.

File: packages/components/src/visually-hidden/index.js [hunk 4]
```diff
@@ -1,46 +1,3 @@
...
-		as,
-		className: classnames( 'components-visually-hidden', className ),
-		...props,
-		ref: forwardedRef,
-	} );
-}
-
-export default withNextComponent( forwardRef( VisuallyHidden ) );
+export { default as VisuallyHidden } from './component';
+export { default } from './component';
```
Comment (youknowriad) [inline packages/components/src/visually-hidden/index.js:1]: Why do we export the component twice?

File: packages/components/src/ui/form-group/form-group-label.js [hunk 5]
```diff
@@ -18,7 +18,7 @@ function FormGroupLabel( { children, id, labelHidden = false, ...props } ) {
 
 	if ( labelHidden ) {
 		return (
-			<VisuallyHidden as="label" htmlFor={ id }>
+			<VisuallyHidden as="label" htmlFor={ id?.toString() }>
```
Comment (youknowriad) [inline packages/components/src/ui/form-group/form-group-label.js:21]: Curious why this change was needed?

Comment (sarayourfriend) [reply packages/components/src/ui/form-group/form-group-label.js:21]: The previous prop types were not comprehensive so they didn't have strict checking of htmlFor. But `id` comes in as either a `number | string` and `htmlFor`, when properly type checked against a `label` only accepts strings.

The polymorphism that existed in the previous component implementation didn't actually make any difference to the types, but now that it's a true `PolymorphicComponent` that has `ViewOwnProps` for its arguments, it actually gets the correct full prop types when using the `as` prop.

File: packages/components/src/visually-hidden/index.js (hunk 4 above)
Comment (sarayourfriend) [reply packages/components/src/visually-hidden/index.js:1]: Some places were importing it as the default and others were importing it by name, so I figured this would minimize the number of changes elsewhere in the components package. I can switch to just one or the other if you prefer though (probably just the named export is best?)

File: packages/components/src/ui/form-group/form-group-label.js (hunk 5 above)
Comment (youknowriad) [reply packages/components/src/ui/form-group/form-group-label.js:21]: Question though, it's just for typing right? I mean, a user passing a numeric value there will still work? (thinking about BC)

File: packages/components/src/visually-hidden/index.js (hunk 4 above)
Comment (youknowriad) [reply packages/components/src/visually-hidden/index.js:1]: I think we historically started with default exports for components. Even if I personally would prefer named components now, I think it's probably better to stay consistent and use default export here.

File: packages/components/src/visually-hidden/component.js (hunk 1 above)
Comment (sarayourfriend) [reply packages/components/src/visually-hidden/component.js:21]: > It looks like a hack to the hooks system :P

That's fair, they probably are 😅

> Is it a requirement to pass all props through a hook?

No, but it's just the way that the g2 components were designed from a technical perspective, to encapsulate all, or as much as possible, logic into a hook that can be reused by other components.

For example, if another component wanted to implement the same logic as `useText` without rendering a `Text` element (like `useHeading` for example, does everything a `Text` can do but with some extras, same with VStack, it's just HStack with some extra props, which in turn is just Flex configured a specific way), this is now possible using these "hacky" hooks.

I'm not saying it's not complicated, but I think approach the technical design of components in this way does add some value around composibility within the library itself (without necessarily exposing all the details of that to consumers of the library).

> Do all components need to support "as" and other features proposed by the "View" component

I think so. It's hard to decide what components _don't_ benefit from the `as` prop. It's an extremely powerful tool for composition.

Icon itself probably doesn't need it, you're correct, but I think the exceptions are few.

> Do all components need to connect to the context system

You'd be surprised. We even have examples of `Text` being accessed via the context system.

I have wondered about the increased overhead of the context system but I think ultimately it's negligible. There were other things like memoization that apparently yielded more significant performance benefits: https://github.com/ItsJonQ/g2/issues/57

File: packages/components/src/ui/form-group/form-group-label.js (hunk 5 above)
Comment (sarayourfriend) [reply packages/components/src/ui/form-group/form-group-label.js:21]: Yup! It'll still work, just a type issue.

**Review by youknowriad — APPROVED** [review]
There's a number of snapshots to remove/update otherwise, it looks good.
